doc/crm/Generic_SAI_Extensions_CRM.md

```diff
@@ -0,0 +1,70 @@
+# Generic SAI Extension Critical Resource Monitoring (CRM)
+
+## Table of Content
+- [Introduction](#Itroduction)
+- [Scope](#Scope)
+- [Requirements](#Requirements)
+- [Highlevel Design](#Highlevel-Design)
+- [Unit Test Automation](#Unit-Test-Automation)
+
+## Revision
+| Rev  |   Date    |                        Author                        | Change Description                 |
+| :--: | :-------: | :--------------------------------------------------: | :--------------------------------: |
+| 0.1  | 1/17/2023 | Intel                                                |  Initial Revision                  |
+
+## Introduction <a name="Introduction"></a>
+High Level Design for critical resource monitoring (crm) in SONiC for Generic SAI Extension tables.
+
+## Scope <a name="Scope"></a>
+The scope of the design is to publish CRM counts, for Generic SAI Extension tables, into the COUNTERS-DB. Support for existing generic crm configurations that is applicable to all resources, like polling interval, is in the scope of this document. No additional configuration specific to extension tables resource is in the scope.
+
+## Requirements <a name="Requirements"></a>
+- Support reporting of free counts
+- Support reporting of available counts
+- Support default watermark checks
+
+## Highlevel Design <a name="Highlevel-Design"></a>
+New crm resource type enum - CrmResourceType::CRM_EXT_TABLE
+  - Each extension table specific stats are stored in,
+    m_resourcesMap.at(CrmResourceType::CRM_EXT_TABLE).countersMap[EXT_TABLE_STATS:"table-name"]
+
+All existing generic crm configuration, like polling-interval, is applicable to this new crm type.
+
+Currently supported default watermark check for other existing crm resource types is also applicable to every extension table under this new crm type.
+
+
+### API to retrieve available counts
+sai_object_type_get_availability() is an existing SAI API used to query any resource specific availability count. The same API is to be used to query generic SAI extension table resource availability as well. API call is to have following object-type and attributes as parameters,
+
+  - object-type    : SAI_OBJECT_TYPE_GENERIC_PROGRAMMABLE
+  - attribute-id   : SAI_GENERIC_PROGRAMMABLE_ATTR_OBJECT_NAME
+
+respective data-plane implementing this API should return resource count available for the corresponding extension table matching the generic programmable object name
+
+### APIs for used counter book-keeping
+- CrmOrch::incCrmExtTableUsedCounter()
+- CrmOrch::decCrmExtTableUsedCounter()
+
+P4Orch extension table manager is responsible to call APIs when a specific entry for extension table is successfully created or removed.

```

Comment:
How is the default watermark handled in SONiC currently? Are there any high & low watermarks available with extension tables ?

Reply:
there is a default for low and high watermarks which is associated with every resource type